### PR TITLE
[aes] Use fresh, random intermediate masks in "noreuse" Canright S-Box

### PR DESCRIPTION
**Pull request by vogelpi**
Formal analysis using the REBECCA tool revealed that it is not secure to re-use bits of the input and output masks for these intermediate masks. Using fresh randomness instead allows the "noreuse" version to pass static verification.

This also requires changes in the masking PRNG and the distribution of the randomness inside the cipher core. To simplify future changes to the masking scheme, the PRNG output now consists of a fixed-width part (8 bits output mask per S-Box) and a parameterized part per S-Box, used e.g. for intermediate masks.

Notes:
- ~~This PR depends on #3851 . Please just review the second commit.~~
- I will enable the setting of LFSR seeds and permutations via topgen into AES in a separate PR as it touches many files. I didn't want to further complicate this PR.

### CODE REVIEW

**Comment by vogelpi** [comment]
I would like to push this one forward. If you hide the whitespace changes, the number of affected lines is only around 200. Please take a look when you get a chance @tjaychen, @eunchan , @cdgori , @msfschaffner .

File: hw/ip/aes/rtl/aes_pkg.sv
Comment (msfschaffner) [inline hw/ip/aes/rtl/aes_pkg.sv:240]: nit: this may need some spaces around ops (also further below)

**Review by msfschaffner — APPROVED** [review]
RTL updates LGTM. You may also want to get an ack from @cdgori, since I did not dive too much into algorithmic details.

File: hw/ip/aes/rtl/aes_pkg.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_pkg.sv:240]: I am already using the same style all over AES and it's lint clean. So I would prefer to change it in a separate PR, for the whole design. Do you think it would help readability?

**Comment by vogelpi** [comment]
Thanks @msfschaffner for taking a look, that's very much appreciated!

**Comment by vogelpi** [comment]
Update: rebased on top of master to fix a conflict. I am merging this now as I need to add other changes for the DOM S-Box. If there should come more security feedback, I will integrate that in a follow-up PR.

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:158]: Sorry for comment-after-merge, but maybe you want to clarify somehow right here that t=q doesn't work?  perhaps "Canright suggests to use t = q but further analysis shows this is not sufficient, see below" ?

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:163]: I believe a similar clarification is needed for s=m ?

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:239]: This comment is a little confusing (or even misleading?) now.  In the previous implementation it made sense when you had to perform the mask switch from m to n for a_inv.  I might just say:

// Note: a_inv is masked by s (= n), a is masked by m

WDYT?

File: hw/ip/aes/rtl/aes_cipher_core.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_cipher_core.sv:270]: I did not try to manually unroll all these part-select values to check everything - but, I assume that if you got this wrong and overlapped some regions, REBECCA would fail (or the compile would fail).

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:14]: nit: this -> these

**Review by cdgori — COMMENTED** [review]
Suggestions for clarifications in the comments you updated - no functional changes.

File: hw/ip/aes/rtl/aes_cipher_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_core.sv:270]: That's a good point! REBECCA doesn't catch that because it's working on and individual S-Box only. At this point we already assume that all mask bits are independent.

But I would have assumed that in case of some overlap, lint would complain about the unused parts. However, it doesn't. Maybe because of the function calls, I am not sure. Need to further investigate.

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:158]: No worries. That sounds good to me.

File: hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_sbox_canright_masked_noreuse.sv:239]: Yes, you are right. I have changed the comment as suggested.

**Review by vogelpi — COMMENTED** [review]
Thanks @cdgori for your feedback! I made another PR to clarify the comments here: #4493 

The thing with the mask overlap is tricky. We should have a reliable way (compile or lint errors) to catch such cases. I need to further look into that.

File: hw/ip/aes/rtl/aes_cipher_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_core.sv:270]: Update: I've implemented a check by re-assembling the PRNG output from the masks and randomness provided to the S-Boxes and then comparing these to signals with an assertion here #4494.

**Pull request merged**
